runner/src/runner.rs

```diff
@@ -0,0 +1,1127 @@
+	opts: Opts,
+}
+
+impl Runner {
+	pub fn new(subxt_api: OnlineClient<PolkadotConfig>, opts: Opts) -> Self {
+		Self { subxt_api, child_proc: None, downloaded_release: None, opts }
+	}
+
+	fn try_load_downloaded_binary(
+		runner: &mut impl RunnerExt,
+		release: &ClientRelease,
+	) -> Result<(), Error> {
+		let (bin_name, bin_path) = runner.get_bin_path(&release.uri)?;
+		let file_content = fs::read(bin_path.clone()).map_err(|_| Error::NoDownloadedRelease)?;
+		let checksum = H256::from_slice(&sha256sum(&file_content));
+
+		if checksum != release.checksum {
+			return Err(Error::IncorrectChecksum)
+		}
+
+		let downloaded_release = DownloadedRelease { checksum, path: bin_path.clone(), bin_name };
+
+		runner.set_downloaded_release(Some(downloaded_release));
+		if let Ok(stringified_path) = bin_path.into_os_string().into_string() {
+			log::info!(
+				"Loaded binary from disk, at {}, with checksum {:#x}",
+				stringified_path,
+				checksum
+			);
+		}
+		Ok(())
+	}
+
+	async fn download_binary(
+		runner: &mut impl RunnerExt,
+		release: ClientRelease,
+	) -> Result<DownloadedRelease, Error> {
+		//recheck checksum to see if binary is the one we have
+		if let Some(downloaded_release) = runner.downloaded_release() {
+			if downloaded_release.checksum.eq(&release.checksum) {
+				log::info!("The on-chain release is already downloaded, skipping re-download");
+				return Ok(downloaded_release.clone())
+			}
+		}
+
+		let (bin_name, bin_path) = runner.get_bin_path(&release.uri)?;
+		log::info!("Downloading {} at: {:?}", bin_name, bin_path);
+
+		// if not, try get the binary
+		let _ = retry_with_log_async(
+			|| {
+				Runner::do_download_binary(bin_path.clone(), release.clone())
+					.into_future()
+					.boxed()
+			},
+			"Error fetching executable".to_string(),
+		)
+		.await?;
+		//Runner::do_download_binary(runner ,release.clone()).await?;
+
+		let downloaded_release =
+			DownloadedRelease { checksum: release.checksum, path: bin_path, bin_name };
+		runner.set_downloaded_release(Some(downloaded_release.clone()));
+		Ok(downloaded_release)
+	}
+
+	async fn do_download_binary(
+		bin_path: PathBuf,
+		release: ClientRelease,
+	) -> Result<&'static str, Error> {
+		let mut file = OpenOptions::new()
+			.read(true)
+			.write(true)
+			// Make the binary executable.
+			// The set permissions are: -rwx------
+			.mode(0o700)
+			.create(true)
+			.open(bin_path.clone())?;
+
+		let bytes = retry_with_log_async(
+			|| Runner::get_request_bytes(release.uri.clone()).into_future().boxed(),
+			"Error fetching executable".to_string(),
+		)
+		.await?;
+
+		file.write_all(&bytes)?;
+		file.sync_all()?;
+
+		//checksum check
+		file.seek(SeekFrom::Start(0))?;
+
+		// Read the content of the file
+		let mut file_content = Vec::new();
+		file.read_to_end(&mut file_content)?;
+
+		let checksum = H256::from_slice(&sha256sum(&file_content));
+
+		if checksum != release.checksum {
+			return Err(Error::IncorrectChecksum)
+		}
+		Ok("success")
```

Comment:
If we don't care about the content of the Result, let's just go for a `Result<(), Error>` and remove the `"success"`